src/rabbit_msg_store.erl

```diff
@@ -1810,15 +1789,61 @@ build_index_worker(Gatherer, State = #msstate { dir = Dir },
             [F|_] -> {F, FileSize}
         end,
     ok = gatherer:in(Gatherer, #file_summary {
-                       file             = File,
-                       valid_total_size = ValidTotalSize,
-                       left             = Left,
-                       right            = Right,
-                       file_size        = FileSize1,
-                       locked           = false,
-                       readers          = 0 }),
+                                  file             = File,
+                                  valid_total_size = ValidTotalSize,
+                                  left             = Left,
+                                  right            = Right,
+                                  file_size        = FileSize1,
+                                  locked           = false,
+                                  readers          = 0 }),
     ok = gatherer:finish(Gatherer).
 
+enqueue_build_index_workers(Gatherer, Left, [], State) ->
+    exit(normal);
+enqueue_build_index_workers(Gatherer, Left, [File|Files], State) ->
+    ok = worker_pool:submit(
```

Comment:
Here I believe the spawned process should queue work into the worker pool only as workers are free, avoiding the massive backlog which is the source of the massive memory usage. In my tests, however, CPU usage hovered around 100% (1 CPU) during the index building. This may be because the worker tasks are quick enough that this single process cannot feed the pool fast enough.

Without theses changes, CPU hovered around 200%, and startup was about 40% faster, so this is not ideal. Ideally we could bound the backlog of the worker pool, such that even a `submit_async` is blocking given that depth (which could default to `infinity` to preserve the existing behavior). Another option might be to look at refactoring the worker pool such that it doesn't degrade for large backlogs (if such a change is possible within acceptable tradeoffs).

Reply:
> the spawned process should queue work into the worker pool only as workers are free

turned out not to be true for worker_pool:submit, but it corrected in a new commit.